### PR TITLE
Improve test coverage

### PR DESCRIPTION
**Pull request by mattgreenham**

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: tests/Controller/AuthorIndexControllerTest.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+use App\BlogsService\Service\TagService;
+use Tests\App\BaseWebTestCase;
+use Tests\App\Builders\AuthorBuilder;
+use Tests\App\Builders\BlogBuilder;
+use Tests\App\Builders\PostBuilder;
+
+/**
+ * @covers \App\Controller\AuthorIndexController
+ */
+class AuthorIndexControllerTest extends BaseWebTestCase
```
Comment (votemike) [inline tests/Controller/AuthorIndexControllerTest.php]: What happens if an Author has no posts. Does this need to be a test case?

File: tests/Controller/AuthorIndexControllerTest.php [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            ->willReturn([]);
+
+        $client->getContainer()->set(PostService::class, $postService);
+
+        $tagService = $this->createMock(TagService::class);
+        $client->getContainer()->set(TagService::class, $tagService);
+
+        $crawler = $client->request('GET', '/blogs/testblog/authors');
+
+        $this->assertEquals('There are no results', $crawler->filterXPath('//h1')->last()->text());
```
Comment (votemike) [inline tests/Controller/AuthorIndexControllerTest.php]: It's grim (in a semantic web kind of way) that this is an H1. Could you create a low priority ticket to audit the blogs site to make sure it is semantically correct. Maybe include this as an example.

File: tests/Controller/HomeControllerTest.php [hunk 3]
```diff
@@ -109,7 +108,7 @@ public function testNoNonArchivedBlogs()
     {
         $client = static::createClient();
 
-        $blog = $this->createBlog('First Blog', true);
+        $blog = BlogBuilder::default()->withName('First Blog')->withIsArchived(true)->build();
```
Comment (votemike) [inline tests/Controller/HomeControllerTest.php]: Do you still need to specify that the name of this blog is "First Blog"?

File: tests/Controller/AuthorIndexControllerTest.php (hunk 1 above)
Comment (mattgreenham) [reply tests/Controller/AuthorIndexControllerTest.php]: Added

File: tests/Controller/AuthorIndexControllerTest.php (hunk 2 above)
Comment (mattgreenham) [reply tests/Controller/AuthorIndexControllerTest.php]: BLOGS-1339 created

File: tests/BaseWebTestCase.php [hunk 4]
```diff
@@ -37,6 +47,17 @@ public function assertHasRequiredResponseHeaders($client, $cacheControl = 'max-a
         }
     }
 
+    public function setUp()
+    {
+        $this->client = static::createClient();
```
Comment (votemike) [inline tests/BaseWebTestCase.php]: Is a client used in *every* test that extends `BaseWebTestCase`?

File: tests/BaseWebTestCase.php [hunk 5]
```diff
@@ -52,4 +73,29 @@ public function extractIstatsLabels(Crawler $crawler): array
 
         return $labels;
     }
+
+    protected function setTestBlog()
+    {
+        $blog = BlogBuilder::default()->build();
```
Comment (votemike) [inline tests/BaseWebTestCase.php]: If you're not setting anything, could this just be a Mock Blog?

File: tests/Controller/AuthorIndexControllerTest.php [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+                'alicesmith' => new iSiteResult(1, 1, 4, [PostBuilder::default()->withTitle('Alice\'s Post')->build()]),
+                'bobjones' => new iSiteResult(1, 1, 4, [PostBuilder::default()->withTitle('Bob\'s Post')->build()]),
+            ]
+        );
+
+        $authorList = $crawler->filterXPath('//ol//li');
+        $this->assertEquals(2, $authorList->count());
+
+        $firstAuthor = $authorList->first();
+        $firstAuthorAuthor = $firstAuthor->filterXPath('//div[@class="grid 1/3@bpw"]');
```
Comment (votemike) [inline tests/Controller/AuthorIndexControllerTest.php]: Not the biggest fan of using these layout classes to select the content in the page. If we change the design, that breaks these tests, which should only be testing the content, not the design.... Although I can't think of a better way off the top of my head. I don't know if anyone else has a good idea.

File: tests/Controller/AuthorShowControllerTest.php [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    public function testAuthorWithPosts()
+    {
+        $crawler = $this->getCrawlerForPage(
+            AuthorBuilder::default()
+                ->withName('John Smith')
+                ->build(),
+            [
+                PostBuilder::default()->build(),
```
Comment (votemike) [inline tests/Controller/AuthorShowControllerTest.php]: Can you use mocks if you're not setting anything explicit?

File: tests/BaseWebTestCase.php (hunk 4 above)
Comment (mattgreenham) [reply tests/BaseWebTestCase.php]: Yes

File: tests/Controller/AuthorShowControllerTest.php (hunk 7 above)
Comment (mattgreenham) [reply tests/Controller/AuthorShowControllerTest.php]: In this case, the posts need to exist as we're asserting they're included in the page - have changed the ones in BaseWebTestCase, though.

File: tests/BaseWebTestCase.php [hunk 8]
```diff
@@ -2,11 +2,22 @@
...
+use App\BlogsService\Domain\Blog;
+use App\BlogsService\Domain\Tag;
+use App\BlogsService\Infrastructure\IsiteResult;
+use App\BlogsService\Service\BlogService;
+use App\BlogsService\Service\TagService;
+use App\Helper\ApplicationTimeProvider;
+use Symfony\Bundle\FrameworkBundle\Client;
 use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
 use Symfony\Component\DomCrawler\Crawler;
+use Tests\App\Builders\TagBuilder;
```
Comment (votemike) [inline tests/BaseWebTestCase.php]: Is this used? Should you run an Optimize Imports over this branch?

**Pull request merged**